GH-10083: Migrate `spring-integration-ws` module to Jspecify

anthologia opened the pull request:
Related to: https://github.com/spring-projects/spring-integration/issues/10083
- Replace `org.springframework.lang.Nullable` with `org.jspecify.annotations.Nullable`
- Migrate `package-info.java` files to use `@NullMarked` annotation
- Add `@SuppressWarnings("NullAway.Init")` for fields initialized in lifecycle methods

For the array fields `BaseWsOutboundGatewaySpec.messageSenders` and `BaseWsOutboundGatewaySpec.gatewayInterceptors`, I applied `@SuppressWarnings("NullAway.Init")` instead of `@Nullable` because adding `@Nullable` still produces the same NullAway warnings.

anthologia committed:
GH-10083: Migrate `spring-integration-ws` module to Jspecify

Related to: https://github.com/spring-projects/spring-integration/issues/10083

- Replace `org.springframework.lang.Nullable` with `org.jspecify.annotations.Nullable`
- Migrate `package-info.java` files to use `@NullMarked` annotation
- Add `@SuppressWarnings("NullAway.Init")` for fields initialized in lifecycle methods

Signed-off-by: Jooyoung Pyoung <pyoungjy@gmail.com>

anthologia commented:
I found some packages don't have `@org.jspecify.annotations.NullMarked` applied.. I'll add them in a follow-up commit !

artembilan commented:
> I'll add them in a follow-up commit !

Right. Let's finish with the current change you have so far and then we will go from there for others.
If that is more convenient for you, we may even go separate PRs.

Thanks

anthologia committed:
Apply JSpecify null safety annotation improvements

- Use proper array nullability: `WebServiceMessageSender @Nullable []`
- Fix inner class annotation placement: `DefaultUriBuilderFactory.@Nullable EncodingMode`
- Set default Jaxb2Marshaller for gatewayMarshaller field

Signed-off-by: Jooyoung Pyoung <pyoungjy@gmail.com>

anthologia commented:
I've applied all the requested changes!
- Set `gatewayMarshaller` to use `Jaxb2Marshaller` as default since it's an implementation of `Marshaller`. Thanks!
- Updated all `@Nullable` annotations to be placed between access modifier and type for consistency.
  - Should we standardize `@Nullable` placement convention project-wide? 🤔 Currently there seems to be mixed usage.
  
I'll add commits for other packages in the `ws` module to this PR rather than separate PRs! Is that fine?

artembilan commented:
> Currently there seems to be mixed usage.

I know, but the easier way to proceed is to replace imports and leave as is.
We can clean up the proper placement as we will go eventually.
If that is fine for you, feel free to make it as it is recommended by JSpecify, but I'm not picky on that.

It is fine to continue in this PR.

Thanks.

artembilan commented on spring-integration-ws/src/main/java/org/springframework/integration/ws/dsl/MarshallingWsOutboundGatewaySpec.java:
```diff
@@ -63,9 +66,9 @@ public static class MarshallingWsOutboundGatewayNoTemplateSpec
 			MarshallingWebServiceOutboundGateway> {
 
-		protected Marshaller gatewayMarshaller; // NOSONAR
+		protected Marshaller gatewayMarshaller = new Jaxb2Marshaller(); // NOSONAR
```
Therefore we can remove that noisy `// NOSONAR`.

Honestly we discontinued to use SonarQube in our projects.
So, all those comments are redundant from now on.

Thanks

anthologia committed:
Apply Nullability to remaining `spring-integration-ws` module packages

* Add `@NullMarked` to all the `ws` packages
* Add `@Nullable` or `@SuppressWarnings("NullAway.Init")` whenever it is requested
* Add defensive null checks for better safety (`DefaultSoapHeaderMapper`, `SimpleWebServiceOutboundGateway`)

Signed-off-by: Jooyoung Pyoung <pyoungjy@gmail.com>

artembilan commented on spring-integration-ws/src/main/java/org/springframework/integration/ws/AbstractWebServiceOutboundGateway.java:
```diff
@@ -84,15 +87,17 @@ public abstract class AbstractWebServiceOutboundGateway extends AbstractReplyPro
 
-	public AbstractWebServiceOutboundGateway(final String uri, WebServiceMessageFactory messageFactory) {
+	@SuppressWarnings("NullAway")
+	public AbstractWebServiceOutboundGateway(final String uri, @Nullable WebServiceMessageFactory messageFactory) {
```
This is not correct. 
The `WebServiceTemplate` expects from us a `WebServiceMessageFactory` as a not null.
If you see somewhere it is, let's mitigate it over there!

anthologia replied:
then.. when ctors pass null for `WebServiceMessageFactory`, I'll replace null with `SaajSoapMessageFactory` as the default. Does this sound good?

Based on the [WebServiceTemplate.properties](https://github.com/spring-projects/spring-ws/blob/main/spring-ws-core/src/main/resources/org/springframework/ws/client/core/WebServiceTemplate.properties), `SaajSoapMessageFactory` is indeed the default strategy.

artembilan replied:
Why would we allow to pass null at all?
The `WebServiceTemplate` does not expect that from us, so let's comeback to not-null for this ctors!

anthologia replied:
Right, I meant that I'll remove `@Nullable` from the constructor parameters and fix the calling sites to pass `SaajSoapMessageFactory` instance instead of null.

For example:
```java
// Instead of
super(uri, null);

// I'll change to  
super(uri, new SaajSoapMessageFactory());

Am I still misunderstanding something?

artembilan replied:
I see:
```
	public MarshallingWebServiceOutboundGateway(DestinationProvider destinationProvider,
			WebServiceTemplate webServiceTemplate) {
		super(destinationProvider, null);
		doSetWebServiceTemplate(webServiceTemplate);
	}
```
I think it is better to introduce a `protected` ctor in the `AbstractWebServiceOutboundGateway` based on the `WebServiceTemplate` then.

artembilan commented on spring-integration-ws/src/main/java/org/springframework/integration/ws/SimpleWebServiceOutboundGateway.java:
```diff
@@ -68,7 +69,7 @@ public SimpleWebServiceOutboundGateway(DestinationProvider destinationProvider,
 	public SimpleWebServiceOutboundGateway(DestinationProvider destinationProvider,
 			@Nullable SourceExtractor<?> sourceExtractor,
-			WebServiceMessageFactory messageFactory) {
+			@Nullable WebServiceMessageFactory messageFactory) {
```
We have to fix those places to not provide a null for the `WebServiceMessageFactory`.

artembilan commented on spring-integration-ws/src/main/java/org/springframework/integration/ws/MarshallingWebServiceOutboundGateway.java:
```diff
@@ -45,11 +46,12 @@ public class MarshallingWebServiceOutboundGateway extends AbstractWebServiceOutb
 		configureMarshallers(marshaller, unmarshaller);
 	}
 
+	@SuppressWarnings("NullAway")
```
I'm not very OK with these `@SuppressWarnings("NullAway")`, but we can look into that separately if that is too much for this change cycle.

artembilan commented:
Hey @anthologia !

Any updates, please?

anthologia committed:
Fix WebServiceMessageFactory null handling

* Add constructor overloads without messageFactory parameter
* Clean up unnecessary @SuppressWarnings("NullAway") annotations

Signed-off-by: Jooyoung Pyoung <pyoungjy@gmail.com>

artembilan commented:
The logic there is like this:
```
	/** Creates a new {@code WebServiceTemplate} using default settings. */
	public WebServiceTemplate() {
		initDefaultStrategies();
	}

	/**
	 * Creates a new {@code WebServiceTemplate} based on the given message factory.
	 * @param messageFactory the message factory to use
	 */
	public WebServiceTemplate(WebServiceMessageFactory messageFactory) {
		setMessageFactory(messageFactory);
		initDefaultStrategies();
	}
```

So, the contract enforces us to provide not-null if we chose that specific ctor.

You may raise a GH issue in [Spring WS](https://github.com/spring-projects/spring-ws) if you still think that it is OK to have it nullable in the second ctor since the first one simply assumes that it really can be null.

artembilan reviewed (changes requested):
Thank you for an update.
Please, give me some feedback on those comments I have left.
As a fix or reject with respective arguments.

anthologia commented on spring-integration-ws/src/main/java/org/springframework/integration/ws/dsl/BaseWsOutboundGatewaySpec.java:
```diff
@@ -51,27 +53,30 @@ public abstract class BaseWsOutboundGatewaySpec<
 	protected String uri; // NOSONAR
 
+	@SuppressWarnings("NullAway.Init")
 	protected WebServiceMessageFactory webServiceMessageFactory; // NOSONAR
```
I agree that `webServiceMessageFactory` should be `@Nullable`.

But I'm not sure if `template` should be `@Nullable`. Because excluding the `*NoTemplateSpec`, two meaningful impls (`SimpleWsOutboundGatewaySpec` and `MarshallingWsOutboundGatewaySpec`) both receive `template` as a non-null param in their ctors.

And, if `template` becomes `@Nullable`, the setter would end up looking like this, which feels awkward to me. 😞

```java
protected final void doSetWebServiceTemplate(@Nullable WebServiceTemplate template) {
    Assert.notNull(template, "'webServiceTemplate' must not be null");
    this.webServiceTemplate = template;
    this.webServiceTemplateExplicitlySet = true;
}
```

artembilan replied:
Sorry I meant like this:
```
protected final WebServiceTemplate template;

protected @Nullable String uri;

protected @Nullable WebServiceMessageFactory webServiceMessageFactory;
```

Therefore `template` is not null as you've just explained.
And that is exactly what I meant with original comment on this line.

anthologia replied:
Ah, got it! That makes perfect sense. Thank you.

anthologia committed:
Apply review feedback to simplify nullable handling and clean up suppressions

* Use `@Nullable` `WebServiceMessageFactory` with conditional `webServiceTemplate` creation
* Mark `uri` and webServiceMessageFactory fields as `@Nullable`, keep template as final
* Remove all `@SuppressWarnings("NullAway")` annotations
* Remove unnecessary ctor overloads from previous approach

Signed-off-by: Jooyoung Pyoung <pyoungjy@gmail.com>
